usr/lib/common/mech_openssl.c

```diff
@@ -4983,3 +4994,388 @@ CK_RV openssl_specific_hmac_final(SIGN_VERIFY_CONTEXT *ctx, CK_BYTE *signature,
+
+    /* Get prime2 a BIGNUM */
+    rc = template_attribute_get_non_empty(key_obj->template, CKA_PRIME_2,
+                                          &prime2);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to get CKA_PRIME_2\n");
+        goto done;
+    }
+
+    q = BN_CTX_get(bn_ctx);
+    if (q == NULL ||
+        BN_bin2bn(prime2->pValue, prime2->ulValueLen, q) == NULL) {
+        TRACE_ERROR("BN_CTX_get/BN_bin2bn failed for prime2\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+    BN_set_flags(q, BN_FLG_CONSTTIME);
+
+    d = BN_CTX_get(bn_ctx);
+    if (d == NULL) {
+        TRACE_ERROR("BN_CTX_get failed to get d\n");
+        rc =  CKR_FUNCTION_FAILED;
+        goto done;
+    }
+    BN_set_flags(d, BN_FLG_CONSTTIME);
+
+    /*
+     * phi(n) = (p - 1 )(q - 1) = n - p - q + 1
+     * d = e ^{-1} mod phi(n).
+     */
+    if (BN_copy(d, n) == NULL ||
+        BN_sub(d, d, p) == 0 ||
+        BN_sub(d, d, q) == 0 ||
+        BN_add_word(d, 1) == 0 ||
+        BN_mod_inverse(d, e, d, bn_ctx) == NULL) {
+        TRACE_ERROR("Failed to calculate private key part d\n");
+        rc =  CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    if (BN_bn2binpad(d, priv_exp, priv_exp_len) <= 0) {
+        TRACE_ERROR("BN_bn2binpad failed\n");
+        rc =  CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+done:
+    BN_CTX_free(bn_ctx);
+
+    return rc;
+}
+
+CK_RV openssl_specific_rsa_derive_kdk(STDLL_TokData_t *tokdata, OBJECT *key_obj,
+                                      const CK_BYTE *in, CK_ULONG inlen,
+                                      CK_BYTE *kdk, CK_ULONG kdklen)
+{
+    CK_ATTRIBUTE *priv_exp_attr = NULL, *modulus = NULL;
+    CK_BYTE *priv_exp = NULL, *buf = NULL;
+    EVP_PKEY *pkey = NULL;
+    EVP_MD_CTX *mdctx = NULL;
+    const EVP_MD *md = NULL;
+    size_t md_len;
+    unsigned char d_hash[SHA256_HASH_SIZE] = { 0 };
+    CK_RV rc;
+
+    /*
+     * The implementation of this function is copied from OpenSSL's function
+     * derive_kdk() in crypto/rsa/rsa_ossl.c and is slightly modified to fit to
+     * the OpenCryptoki environment.
+     * Changes include:
+     * - Different variable and define names.
+     * - Usage of TRACE_ERROR to report errors and issue debug messages.
+     * - Different return codes.
+     * - Different code to get the private key component 'd'.
+     * - Use of the EVP APIs instead of the internal APIs for Digest and HMAC
+     *   operations.
+     */
+
+    if (kdklen != SHA256_HASH_SIZE) {
+        TRACE_ERROR("KDK length is wrong\n");
+        return CKR_ARGUMENTS_BAD;
+    }
+
+    rc = template_attribute_get_non_empty(key_obj->template, CKA_MODULUS,
+                                          &modulus);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to get CKA_MODULUS\n");
+        return rc;
+    }
+
+    buf = calloc(1, modulus->ulValueLen);
+    if (buf == NULL) {
+        TRACE_ERROR("Failed to allocate a buffer for private exponent\n");
```

Comment:
buffer for modulus

Reply:
Ah ok ... it's in fact the private exponent, which has the same length as the modulus ...